ExtractorUtils/BaseExtractor.cs

```diff
@@ -0,0 +1,288 @@
+        /// </summary>
+        protected CogniteDestination Destination { get; }
+        /// <summary>
+        /// Timeseries upload queue
+        /// </summary>
+        protected TimeSeriesUploadQueue TSUploadQueue { get; private set; }
+        /// <summary>
+        /// Event upload queue
+        /// </summary>
+        protected EventUploadQueue EventUploadQueue { get; private set; }
+        /// <summary>
+        /// Raw upload queues, by dbName-tableName and type.
+        /// </summary>
+        protected Dictionary<(string name, Type type), IUploadQueue> RawUploadQueues { get; private set; }
+            = new Dictionary<(string name, Type type), IUploadQueue>();
+        /// <summary>
+        /// Scheduler for running various periodic tasks
+        /// </summary>
+        protected PeriodicScheduler Scheduler { get; private set; }
+        /// <summary>
+        /// Cancellation token source
+        /// </summary>
+        protected CancellationTokenSource Source { get; private set; }
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="config">Configuration object</param>
+        /// <param name="destination">Cognite destination</param>
+        public BaseExtractor(BaseConfig config, CogniteDestination destination)
+        {
+            Config = config;
+            Destination = destination;
+            
+        }
+
+        /// <summary>
+        /// Verify that the extractor is configured correctly.
+        /// </summary>
+        /// <returns>Task</returns>
+        protected virtual async Task TestConfig()
+        {
+            await Destination.TestCogniteConfig(Source.Token).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Method called to start the extractor.
+        /// </summary>
+        /// <returns></returns>
+        public virtual async Task Start(CancellationToken token)
+        {
+            Source?.Dispose();
+            Scheduler?.Dispose();
+            Source = CancellationTokenSource.CreateLinkedTokenSource(token);
+            Scheduler = new PeriodicScheduler(Source.Token);
+            await TestConfig().ConfigureAwait(false);
+            try
+            {
+                await Start().ConfigureAwait(false);
+                await Scheduler.WaitForAll().ConfigureAwait(false);
+            }
+            finally
+            {
+                await OnStop().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Internal method starting the extractor. Should handle any creation of timeseries,
+        /// setup of source systems, and calls to the various Schedule and Create protected methods.
+        /// </summary>
+        /// <returns></returns>
+        protected abstract Task Start();
+
+
+        /// <summary>
+        /// Called when the extractor is stopping.
+        /// </summary>
+        /// <returns></returns>
+        protected virtual Task OnStop()
```

Comment:
Nice - I guess this can be utilized to automate reporting extraction pipeline runs?

Reply:
is there a way to catch if it stopped successfully or not in here?

Reply:
Maybe? I'll have to think about it, ideally it shouldn't require anything from the user, and OnStop is intended as a callback for the extractor developer.

Reply:
I don't think it matters if this fails. If they use the runner method it will be caught there, if not they will have to catch it themselves. Any cleanup the extractor library needs to do is done independent of success or failure here.